src/pages/task/taskTabs/buildBaronAndAnnotations/AddIssueModal/state.ts

```diff
@@ -0,0 +1,90 @@
+import { useReducer } from "react";
+import { useSpruceConfig } from "hooks";
+import { numbers } from "utils";
+import { validateJiraURL } from "utils/validators";
+
+const { toDecimal } = numbers;
+
+interface addIssueState {
+  url: string;
+  issueKey: string;
+  canSubmit: boolean;
+  isURLValid: boolean;
+  isKeyValid: boolean;
+  confidenceScore: string | null;
+}
+
+type Action =
+  | { type: "reset" }
+  | { type: "setUrl"; url: string; jiraURL: string }
+  | { type: "setKey"; issueKey: string }
+  | { type: "setConfidenceScore"; confidenceScore: string | null };
+
+const init = () => ({
+  url: "",
+  issueKey: "",
+  canSubmit: false,
+  isURLValid: false,
+  isKeyValid: false,
+  confidenceScore: null,
+});
+
+const reducer = (state: addIssueState, action: Action) => {
+  switch (action.type) {
+    case "reset":
+      return init();
+    case "setUrl": {
+      const isURLValid = validateJiraURL(action.jiraURL, action.url);
+      return {
+        ...state,
+        url: action.url,
+        canSubmit: isURLValid && state.isKeyValid,
+        isURLValid,
+      };
+    }
+    case "setKey": {
+      const isKeyValid = action.issueKey.length > 0;
+      return {
+        ...state,
+        issueKey: action.issueKey,
+        isKeyValid,
+        canSubmit: isKeyValid && state.isURLValid,
+      };
+    }
+    case "setConfidenceScore": {
+      const isNumber = !Number.isNaN(action.confidenceScore);
+      const isValid =
+        toDecimal(action.confidenceScore) <= 1 &&
+        toDecimal(action.confidenceScore) >= 0;
+      if (isNumber && isValid) {
+        return {
+          ...state,
+          confidenceScore: action.confidenceScore,
+        };
+      }
+      return state;
+    }
+    default:
+      throw new Error("Unrecognized action type");
+  }
+};
+
+export const useAddIssueModal = () => {
+  const spruceConfig = useSpruceConfig();
+  const [state, dispatch] = useReducer(reducer, init());
+
+  const setUrl = (url: string) => {
+    dispatch({ type: "setUrl", url, jiraURL: spruceConfig.jira.host });
+  };
+  const setKey = (issueKey: string) => {
+    dispatch({ type: "setKey", issueKey });
+  };
+  const setConfidenceScore = (confidenceScore: string | null) => {
+    dispatch({ type: "setConfidenceScore", confidenceScore });
+  };
+  const reset = () => {
+    dispatch({ type: "reset" });
+  };
+
+  return [state, { setUrl, setKey, setConfidenceScore, reset }] as const;
```

Comment:
why do we need the type alias?

Reply:
Its a const assertion. https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions it just ensures the values are readonly

Typescript was having difficulty interpreting the type for this array. 
![image](https://user-images.githubusercontent.com/4605522/167026616-0ea7a375-e37e-4900-b931-fefb3ad4ff24.png)